manual/tools/texquote2.ml

```diff
@@ -0,0 +1,143 @@
+type environment =
+  | Caml
+  | Verbatim
+  | Verbatim_like
+
+let (current_environment : environment option ref) = ref None
+
+let in_quotes = ref false
+
+let verbatim_end_in = ref ""
+
+let verbatim_end_out = ref ""
+
+let is_alpha c =
+  ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z')
+
+let is_prefix prefix str =
+  let length_prefix = String.length prefix in
+  let length_str = String.length str in
+  if length_prefix > length_str
+  then false
+  else (String.sub str 0 length_prefix) = prefix
+
+let escape = function
+  | ' ' | '\n' -> "\\ "
+  | '{' -> "{\\char123}"
+  | '}' -> "{\\char125}"
+  | '^' -> "{\\char94}"
+  | '_' -> "{\\char95}"
+  | '\\' -> "{\\char92}"
+  | '~' -> "{\\char126}"
+  | '$' -> "\\$"
+  | '&' -> "{\\char38}"
+  | '#' -> "\\#"
+  | '%' -> "\\%"
+  | '\'' -> "{\\textquotesingle}"
+  | '`' -> "{\\textasciigrave}"
+  | _ -> ""
+
+let process_normal_line line =
+  let (verb_mark : char option ref) = ref None in
+  let l = String.length line in
+  let i = ref 0 in
+  while !i<l do
+    match !verb_mark with
+    | None ->
+      (match line.[!i] with
+      | '"' ->
+        let r = if !in_quotes then "}}" else "{\\machine{" in
+        print_string r;
+        in_quotes := not !in_quotes;
+        incr i;
+      | '\\' ->
+        if !in_quotes
+        then begin
+          if (!i < l-1) && (line.[!i+1] = '"' || line.[!i+1] = '\\')
+          then incr i;
+          let t = escape line.[!i] in
+          if t<>"" then print_string t else print_char line.[!i];
+          incr i;
+        end else if is_prefix "\\verb" (String.sub line !i (l - !i))
+                    && not (is_alpha line.[!i+5])
+        then begin
+          i := !i+5;
+          verb_mark := Some line.[!i];
+          print_string "\\verb";
+          print_char line.[!i];
+          incr i;
+        end else (print_char '\\'; incr i)
+      | _ ->
+        if !in_quotes && (escape line.[!i] <> "")
+        then print_string (escape line.[!i])
+        else print_char line.[!i];
+        incr i;        
+      )
+    | Some mark ->
+      if line.[!i] = mark
+      then verb_mark := None
+      else if line.[!i] = '\'' || line.[!i] = '`'
+      then Printf.eprintf "Warning: %c found in \\verb\n" line.[!i];
+      print_char line.[!i];
+      incr i;        
+  done
+
+let process_line line = function
```

Comment:
I would rewrite this as a function: `string -> environment -> environment` and get rid of the global variable `current_environment`.